src/dialects/abstract/data-types.ts

```diff
@@ -0,0 +1,1826 @@
+    this.fields = fields;
+  }
+}
+
+/**
+ * A virtual value that is not stored in the DB. This could for example be useful if you want to provide a default value in your model that is returned to the user but not stored in the DB.
+ *
+ * You could also use it to validate a value before permuting and storing it. VIRTUAL also takes a return type and dependency fields as arguments
+ * If a virtual attribute is present in `attributes` it will automatically pull in the extra fields as well.
+ * Return type is mostly useful for setups that rely on types like GraphQL.
+ *
+ * @example <caption>Checking password length before hashing it</caption>
+ * sequelize.define('user', {
+ *   password_hash: DataTypes.STRING,
+ *   password: {
+ *     type: DataTypes.VIRTUAL,
+ *     set: function (val) {
+ *        // Remember to set the data value, otherwise it won't be validated
+ *        this.setDataValue('password', val);
+ *        this.setDataValue('password_hash', this.salt + val);
+ *      },
+ *      validate: {
+ *         isLongEnough: function (val) {
+ *           if (val.length < 7) {
+ *             throw new Error("Please choose a longer password")
+ *          }
+ *       }
+ *     }
+ *   }
+ * })
+ *
+ * # In the above code the password is stored plainly in the password field so it can be validated, but is never stored in the DB.
+ *
+ * @example <caption>Virtual with dependency fields</caption>
+ * {
+ *   active: {
+ *     type: new DataTypes.VIRTUAL(DataTypes.BOOLEAN, ['createdAt']),
+ *     get: function() {
+ *       return this.get('createdAt') > Date.now() - (7 * 24 * 60 * 60 * 1000)
+ *     }
+ *   }
+ * }
+ *
+ */
+export type VIRTUAL<T> = _VIRTUAL<T>;
+export const VIRTUAL = classToInvokable(_VIRTUAL);
+
+interface EnumOptions<Member extends string> {
+  values: Member[];
+}
+
+type EnumParams<Member extends string> =
+  | []
+  | [EnumOptions<Member>]
+  | [Member[]]
+  | Array<Member | Member[]>;
+
+class _ENUM<Member extends string> extends ABSTRACT<Member> {
+  public static readonly key = 'ENUM';
+  public values: Member[];
+  protected options: EnumOptions<Member>;
+
+  /**
+   * @param {...any|{ values: any[] }|any[]} args either array of values or options object with values array. It also supports variadic values
+   */
+  constructor(options: EnumOptions<Member>);
+  constructor(members: Member[]);
+  constructor(...members: Array<Member | Member[]>);
+  constructor(...args: EnumParams<Member>); // This is the overload TS will chose when using ConstructorParams and is needed.
+  constructor(...args: EnumParams<Member>) {
+    super();
+    let options: EnumOptions<Member>;
+
+    const [arg0] = args;
+
+    if (typeof arg0 === 'object' && !Array.isArray(arg0)) {
+      options = arg0;
+    } else if (args.length === 1 && Array.isArray(arg0)) {
+      options = { values: arg0 };
+    } else {
+      options = {
+        // this rule is irrelevant - this is using `concat` for cases such as `new Union('a', ['b', 'c]])`
```

Comment:
`DataTypes.ENUM('a', ['b', 'c'])` should throw. We should only support

```
DataTypes.ENUM('value', 'another value')
DataTypes.ENUM(['value', 'another value'])
DataTypes.ENUM({
  values: ['value', 'another value']
})
```

to avoid future accidental breaking changes because a refactor broke something someone accidentally used